registry/etcdv3/registry.go

```diff
@@ -0,0 +1,323 @@
+	configListener *configurationListener
+
+	wg   sync.WaitGroup // wg+done for etcd client restart
+	done chan struct{}
+}
+
+func (r *etcdV3Registry) Client() *etcdv3.Client {
+	return r.client
+}
+func (r *etcdV3Registry) SetClient(client *etcdv3.Client) {
+	r.client = client
+}
+func (r *etcdV3Registry) ClientLock() *sync.Mutex {
+	return &r.cltLock
+}
+func (r *etcdV3Registry) WaitGroup() *sync.WaitGroup {
+	return &r.wg
+}
+func (r *etcdV3Registry) GetDone() chan struct{} {
+	return r.done
+}
+func (r *etcdV3Registry) RestartCallBack() bool {
+
+	services := []common.URL{}
+	for _, confIf := range r.services {
+		services = append(services, confIf)
+	}
+
+	flag := true
+	for _, confIf := range services {
+		err := r.Register(confIf)
+		if err != nil {
+			logger.Errorf("(etcdV3ProviderRegistry)register(conf{%#v}) = error{%#v}",
+				confIf, errors.WithStack(err))
+			flag = false
+			break
+		}
+		logger.Infof("success to re-register service :%v", confIf.Key())
+	}
+	return flag
+}
+
+func newETCDV3Registry(url *common.URL) (registry.Registry, error) {
+
+	timeout, err := time.ParseDuration(url.GetParam(constant.REGISTRY_TIMEOUT_KEY, constant.DEFAULT_REG_TIMEOUT))
+	if err != nil {
+		logger.Errorf("timeout config %v is invalid ,err is %v",
+			url.GetParam(constant.REGISTRY_TIMEOUT_KEY, constant.DEFAULT_REG_TIMEOUT), err.Error())
+		return nil, errors.WithMessagef(err, "new etcd registry(address:%+v)", url.Location)
+	}
+
+	logger.Infof("etcd address is: %v", url.Location)
+	logger.Infof("time-out is: %v", timeout.String())
+
+	r := &etcdV3Registry{
+		URL:      url,
+		birth:    time.Now().UnixNano(),
+		done:     make(chan struct{}),
+		services: make(map[string]common.URL),
+	}
+
+	if err := etcdv3.ValidateClient(r,
+		etcdv3.WithName(etcdv3.RegistryETCDV3Client),
+		etcdv3.WithTimeout(timeout),
+		etcdv3.WithEndpoints(url.Location),
+	); err != nil {
+		return nil, err
+	}
+
+	r.wg.Add(1)
+	go etcdv3.HandleClientRestart(r)
+
+	r.listener = etcdv3.NewEventListener(r.client)
+	r.configListener = NewConfigurationListener(r)
+	r.dataListener = NewRegistryDataListener(r.configListener)
+
+	return r, nil
+}
+
+func (r *etcdV3Registry) GetUrl() common.URL {
+	return *r.URL
+}
+
+func (r *etcdV3Registry) IsAvailable() bool {
+
+	select {
+	case <-r.done:
+		return false
+	default:
+		return true
+	}
+}
+
+func (r *etcdV3Registry) Destroy() {
+
+	if r.configListener != nil {
+		r.configListener.Close()
+	}
+	r.stop()
+}
+
+func (r *etcdV3Registry) stop() {
+
+	close(r.done)
+
+	// close current client
+	r.client.Close()
+
+	r.cltLock.Lock()
+	r.client = nil
+	r.services = nil
+	r.cltLock.Unlock()
+}
+
+func (r *etcdV3Registry) Register(svc common.URL) error {
+
+	role, err := strconv.Atoi(r.URL.GetParam(constant.ROLE_KEY, ""))
+	if err != nil {
+		return errors.WithMessage(err, "get registry role")
+	}
+
+	r.cltLock.Lock()
+	if _, ok := r.services[svc.Key()]; ok {
+		return errors.New(fmt.Sprintf("Path{%s} has been registered", svc.Path))
```

Comment:
deadlock?

Reply:
Yes,  this is a serious bug, should unlock before return.